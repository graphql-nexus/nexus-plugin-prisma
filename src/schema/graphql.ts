//
// GraphQL root types data & helpers
//

export type RootName = 'Query' | 'Mutation' | 'Subscription'

export const rootNames = {
  Query: 'Query',
  Mutation: 'Mutation',
  Subscription: 'Subscription',
} as const

export const rootNameValues: RootName[] = Object.values(rootNames)

export const isRootName = (x: any): x is RootName => rootNameValues.includes(x)

//
// GraphQL scalar types data & helpers
//

export type ScalarName = 'Int' | 'Float' | 'String' | 'ID' | 'Boolean'

export const scalarsNameValues: ScalarName[] = ['Int', 'Float', 'String', 'ID', 'Boolean']

export const isScalarType = (name: string): boolean => scalarsNameValues.includes(name as any)
