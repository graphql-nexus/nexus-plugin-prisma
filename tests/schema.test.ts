import { objectType } from 'nexus'
import { printSchema } from 'graphql'
import { generateSchema } from './__utils'

jest.setTimeout(20000)

test('simple schema', async () => {
  const User = objectType({
    name: 'User',
    definition(t: any) {
      t.model.id()
      t.model.name()
    },
  })
  const datamodel = `
    model User {
      id    Int @id
      name  String
    }
    `

  const schema = await generateSchema(datamodel, [User])

  expect(printSchema(schema)).toMatchSnapshot()
})

test('it exposes only pagination on relations by default', async () => {
  const datamodel = `
    model User {
      id    Int @id
      posts Post[]
    }

    model Post {
      id    Int @id
    }
    `
  const User = objectType({
    name: 'User',
    definition(t: any) {
      t.model.id()
      t.model.posts()
    },
  })
  const Post = objectType({
    name: 'Post',
    definition(t: any) {
      t.model.id()
    },
  })

  const schema = await generateSchema(datamodel, [User, Post])

  expect(printSchema(schema)).toMatchSnapshot()
})

test('it exposes filtering only if filtering: true', async () => {
  const datamodel = `
    model User {
      id    Int @id
      posts Post[]
    }

    model Post {
      id    Int @id
    }
    `
  const User = objectType({
    name: 'User',
    definition(t: any) {
      t.model.id()
      t.model.posts({ filtering: true })
    },
  })
  const Post = objectType({
    name: 'Post',
    definition(t: any) {
      t.model.id()
    },
  })

  const schema = await generateSchema(datamodel, [User, Post])

  expect(printSchema(schema)).toMatchSnapshot()
})

test('it exposes only id filters', async () => {
  const datamodel = `
    model User {
      id    Int @id
      posts Post[]
    }

    model Post {
      id    Int     @id
      name  String
    }
    `
  const User = objectType({
    name: 'User',
    definition(t: any) {
      t.model.id()
      t.model.posts({ filtering: { id: true } })
    },
  })

  const Post = objectType({
    name: 'Post',
    definition(t: any) {
      t.model.id()
      t.model.name()
    },
  })

  const schema = await generateSchema(datamodel, [User, Post])

  expect(printSchema(schema)).toMatchSnapshot()
})

test('it exposes ordering only if ordering: true', async () => {
  const datamodel = `
    model User {
      id    Int @id
      posts Post[]
    }

    model Post {
      id    Int     @id
      name  String
    }
    `
  const User = objectType({
    name: 'User',
    definition(t: any) {
      t.model.id()
      t.model.posts({ ordering: true })
    },
  })

  const Post = objectType({
    name: 'Post',
    definition(t: any) {
      t.model.id()
      t.model.name()
    },
  })

  const schema = await generateSchema(datamodel, [User, Post])

  expect(printSchema(schema)).toMatchSnapshot()
})

test('it exposes id ordering', async () => {
  const datamodel = `
    model User {
      id    Int @id
      posts Post[]
    }

    model Post {
      id    Int     @id
      name  String
    }
    `
  const User = objectType({
    name: 'User',
    definition(t: any) {
      t.model.id()
      t.model.posts({ ordering: { id: true } })
    },
  })

  const Post = objectType({
    name: 'Post',
    definition(t: any) {
      t.model.id()
      t.model.name()
    },
  })

  const schema = await generateSchema(datamodel, [User, Post])

  expect(printSchema(schema)).toMatchSnapshot()
})

test('it exposes findOne and findMany', async () => {
  const datamodel = `
    model User {
      id    Int @id
    }
    `
  const Query = objectType({
    name: 'Query',
    definition(t: any) {
      t.crud.user()
      t.crud.users()
    },
  })
  const User = objectType({
    name: 'User',
    definition(t: any) {
      t.model.id()
    },
  })

  const schema = await generateSchema(datamodel, [User, Query])

  expect(printSchema(schema)).toMatchSnapshot()
})
