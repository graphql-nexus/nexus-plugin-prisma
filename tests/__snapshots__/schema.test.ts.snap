// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it exposes filtering only if filtering: true 1`] = `
"input IntFilter {
  equals: Int
  not: Int
  in: [Int!]
  notIn: [Int!]
  lt: Int
  lte: Int
  gt: Int
  gte: Int
}

type Post {
  id: Int!
}

type Query {
  ok: Boolean!
}

type User {
  id: Int!
  posts(where: UserPostsWhereInput, skip: Int, after: String, before: String, first: Int, last: Int): [Post!]
}

input UserPostsWhereInput {
  id: IntFilter
  AND: [UserPostsWhereInput!]
  OR: [UserPostsWhereInput!]
  NOT: [UserPostsWhereInput!]
  user: UserPostsWhereInput
}
"
`;

exports[`it exposes findOne and findMany 1`] = `
"type Query {
  user(where: UserWhereUniqueInput!): User
  users(skip: Int, after: String, before: String, first: Int, last: Int): [User!]
}

type User {
  id: Int!
}

input UserWhereUniqueInput {
  id: Int
}
"
`;

exports[`it exposes id ordering 1`] = `
"enum OrderByArg {
  asc
  desc
}

type Post {
  id: Int!
  name: String!
}

type Query {
  ok: Boolean!
}

type User {
  id: Int!
  posts(orderBy: UserPostsOrderByInput, skip: Int, after: String, before: String, first: Int, last: Int): [Post!]
}

input UserPostsOrderByInput {
  id: OrderByArg
}
"
`;

exports[`it exposes only id filters 1`] = `
"input IntFilter {
  equals: Int
  not: Int
  in: [Int!]
  notIn: [Int!]
  lt: Int
  lte: Int
  gt: Int
  gte: Int
}

type Post {
  id: Int!
  name: String!
}

type Query {
  ok: Boolean!
}

type User {
  id: Int!
  posts(where: UserPostsWhereInput, skip: Int, after: String, before: String, first: Int, last: Int): [Post!]
}

input UserPostsWhereInput {
  id: IntFilter
}
"
`;

exports[`it exposes only pagination on relations by default 1`] = `
"type Post {
  id: Int!
}

type Query {
  ok: Boolean!
}

type User {
  id: Int!
  posts(skip: Int, after: String, before: String, first: Int, last: Int): [Post!]
}
"
`;

exports[`it exposes ordering only if ordering: true 1`] = `
"enum OrderByArg {
  asc
  desc
}

type Post {
  id: Int!
  name: String!
}

type Query {
  ok: Boolean!
}

type User {
  id: Int!
  posts(orderBy: UserPostsOrderByInput, skip: Int, after: String, before: String, first: Int, last: Int): [Post!]
}

input UserPostsOrderByInput {
  id: OrderByArg
  name: OrderByArg
}
"
`;

exports[`simple schema 1`] = `
"type Query {
  ok: Boolean!
}

type User {
  id: Int!
  name: String!
}
"
`;
